### ISSUE TITLE
Why Flow why not TypeScript

### ISSUE THREAD

**Issue by nainardev** [popperjs/popper-core#1024]
I mean why isn't the whole library in typescript?

**Comment by FezVrasta** [comment]
Because TypeScript is flawed by design and Flow is a stricter type system.

**Comment by pleerock** [comment]
I think you should give a try to TypeScript then you can build your own opinion based on a real comparison of pros and cons. And most people would say TypeScript is a definite winner ;) Awesome package btw

**Comment by FezVrasta** [comment]
I use both TS and Flow, thanks for the advice though.

**Comment by robinmetral** [comment]
I don't endorse the OP's tone and didn't really want to bring this old thread back to life, but I think that this is worth reconsidering.

I don't know flow, and I'm ready to believe that it's technically a better type system than TypeScript. However that's not the only thing we should consider when choosing technologies, especially in open-source projects. Criteria such as community and ecosystem are just as important.

Concretely speaking, flow is a barrier to entry for contributing to popperjs, and I'm willing to bet that the library is used in many more TS than flow projects out there. Switching would surely improve developer experience overall, while encouraging community contributions.

Let's put ourselves in the shoes of a FE developer who's looking at learning a type system today: which one would they choose?

<img width="500" alt="Screenshot 2021-11-23 at 11 53 58" src="https://user-images.githubusercontent.com/35560568/143012382-c0977c18-ae30-4b2c-bf3c-2bd72de24644.png">
<img width="500" alt="Screenshot 2021-11-23 at 11 54 18" src="https://user-images.githubusercontent.com/35560568/143012387-5ce46fb7-9f22-4fca-8a9b-d6e038e2fd31.png">
<img width="1323" alt="Screenshot 2021-11-23 at 11 55 37" src="https://user-images.githubusercontent.com/35560568/143012388-96f00a72-2ef3-439d-914c-ca83e840851d.png">

Are you particularly attached to flow or would you agree to have a discussion about a potential migration?

**Comment by atomiks** [comment]
Flow is pretty similar to TypeScript, especially with this codebase's types (v3's types are even simpler)

TL;DR for contributing here is:

- `$Shape<>` is `Partial<>`
- `{| |}` are exact objects
- Type predicates can be seen in the `instanceOf` file
- Just use `type` keyword instead of `interface`

Add the Flow VS Code extension (and possibly disable TS typechecking for the project) and that's it

**Comment by FezVrasta** [comment]
Maintaining a Flow codebase allows to easily auto generate typescript types. Doing the opposite is quite difficult, because Flow is a way stricter type system

**Comment by robinmetral** [comment]
Okay, good to know, so you're saying that we're sticking with flow here because it's a "2 for the price of 1" scenario—this makes sense, although using popper in TS didn't feel as smooth to me as using a TS-first library like [unified](https://github.com/unifiedjs/unified). But you're right that the types were there and totally usable with a bit of tweaking.

I still think that there would be more community contributions if the codebase was in TS, but it also looks like you're doing a pretty good job at keeping on top of questions/requests (🙌), so maybe we don't need more contributors for the sake of having more contributors.

I wouldn't be surprised if some day we realize that it would make everyone's life easier if we to switched over to TS, but it seems fair to stick with flow right now. Thanks a lot for the extra rationale and thoughts here and thanks for the library 🍿

**Comment by robinmetral** [comment]
Happy to see that Floating UI is no TypeScript 🎉
